js/client/tree.js

```diff
@@ -0,0 +1,188 @@
+/**
+ * @license
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+  * Base class for a view tree on the client, extended by
+  * NodeTree, EventTree and PropsTree
+  */
+
+// TODO - perf work on
+//        https://cs.chromium.org/chromium/src/ui/accessibility/ax_node_data.cc
+// TODO - screen reader access,
+//      - label for whole row aria-level, posinset, setsize, aria-expanded,
+//        aria-owns, no aria-labelledby, no role="treeitem" in treegrid case
+//      - leave off aria-selected compeltely when stuff is not selectable.
+//      - aria-activedescendant point to row
+// TODO - In desktop/app mode, ability to limit to a specific root
+// TODO - P2: Plain text search and regex search can be done via findAll
+// TODO - P2: resizable panels ala https://codepen.io/rstrahl/pen/eJZQej
+// TODO - P2: ability to hide non-user groups of events e.g. noisy mutations?
+// TODO - P2: ability to have new events in log expanded by default?
+// TODO - P2: pick item with mouse, draw bounds rects
+// TODO - P2: In app, ability to choose root, e.g. desktop vs specific window
+// TODO - P3: Choose columns ?
+// TODO - P3: allow function execution with params? Via config file
+// TODO - P3: Highlights don't work unless we knew about nodeChanged node before
+// TODO - P3: remote access
+// TODO - P3: expose internal id's (would need to add field to AutomationNode)
+// TODO - P4: all searches should be done by background script, passing back
+//            keys, perhaps using alterations to the automation API
+// TODO - P4: allow JSON to be constructed via options dialog
+// TODO - P4: put all config options in JSON
+// Qs:
+//   - htmlTag not working
+//   - Why isn't source=user happening for every click event
+//   - textChanged event, why both event and treeChange?
+//   - treeChanged event, how is this different from other tree changes?
+//   - getImageData() crashes on some nodes
+//   - Hit inspect button, show same old inspector (ext)
+//--
+//   - DCHECK failing in ax_enums.cc at end of ToString(AXState)
+//   - In app, for event listening, should we limit it to one window at a time?
+//     Have a window chooser?
```

Comment:
Yes and a hit test function too, I've been implementing full hit test
support in automation recently and it should be ready now